linked_list/implementation.py

```diff
@@ -1,40 +1,235 @@
+        generator_pointer = self.cur_pointer
+        while generator_pointer != None:
+            yield generator_pointer.elem
+            generator_pointer = generator_pointer.next
 
     def __getitem__(self, index):
-        pass
+        for node_index, element in enumerate(self):
+            if node_index == index:
+                return element
 
+    # Needs to abstract self
     def __add__(self, other):
-        pass
-
+        # Check whether self is an empty LL
+        # Check whether other is an empty LL
+        # If either are empty, use the other in place of the other
+        # If both are occupied, point self.end.next to other.start
+       
+        # If neither is occupied node, return an empty node
+        if self.start == None and other.start == None:
+            return self
+        
+        # Get a copy of self so value by reference doesn't overwrite the original value
+        addition_self = deepcopy(self)
+        addition_other = deepcopy(other)
+        
+        if self.start == self.end and self.start == None:
+            addition_self.start = other.start
+            addition_self.end = other.end
+            addition_self.cur_pointer = addition_self.start
+        elif other.start == other.end and other.start == None:
+            addition_other.start = self.start
+            addition_other.end = self.end
+            addition_other.cur_pointer = addition_other.start
+            return addition_other
+        elif len(self) == 1:
+            addition_self.start.next = other.start
+            addition_self.end = other.end
+        elif len(self) > 1:
+            addition_self.end.next = other.start
+            addition_self.end = other.end
+        else:
+            raise IndexError
+        return addition_self
+    
     def __iadd__(self, other):
-        pass
+        # Both LLs are empty, return an empty LL
+        if self.start == None and other.start == None:
+            return self
+        # Check if self is an empty LL
+        elif self.start == self.end and self.start == None:
+            self.start = other.start
+            self.end = other.end
+            self.cur_pointer = self.start #Needed because start is at other.start now
+            return self
+        # Check if other is an empty LL
+        elif other.start == other.end and other.start == None:
+            return self
+        else:
+            self.end.next = other.start
+            self.end = other.end
+            return self
 
     def __eq__(self, other):
-        pass
+        if len(self) != len(other):
+            return False
+        for index, item in enumerate(self):
+            if item != other[index]:
+                return False
+        return True
 
+    def __ne__(self, other):
+        if len(self) != len(other):
+            return True
+        for index, item in enumerate(self):
+            if item != other[index]:
+                return True
+        return False
+            
```

Comment:
When I wrote early on, I didn't realize we can use the magic methods in each other after them being written. In retrospect this would be the smart thing to do.